src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs

```diff
@@ -757,331 +678,167 @@ public virtual void Write(ReadOnlySpan<byte> buffer)
+                cancellationToken.IsCancellationRequested ?
+                    Task.FromCanceled<int>(cancellationToken) :
+                    Task.FromResult(0);
 
-            public override ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default)
-            {
-                return new ValueTask<int>(0);
-            }
+            public override ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken) =>
+                cancellationToken.IsCancellationRequested ?
+                    ValueTask.FromCanceled<int>(cancellationToken) :
+                    default;
 
-            public override int ReadByte()
-            {
-                return -1;
-            }
+            public override int ReadByte() => -1;
 
-            public override void Write(byte[] buffer, int offset, int count)
-            {
-            }
+            public override void Write(byte[] buffer, int offset, int count) { }
 
-            public override void Write(ReadOnlySpan<byte> buffer)
-            {
-            }
+            public override void Write(ReadOnlySpan<byte> buffer) { }
 
-            public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
-            {
-                return cancellationToken.IsCancellationRequested ?
+            public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken) =>
+                cancellationToken.IsCancellationRequested ?
                     Task.FromCanceled(cancellationToken) :
                     Task.CompletedTask;
-            }
 
-            public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default)
-            {
-                return cancellationToken.IsCancellationRequested ?
+            public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default) =>
+                cancellationToken.IsCancellationRequested ?
                     ValueTask.FromCanceled(cancellationToken) :
                     default;
-            }
 
-            public override void WriteByte(byte value)
-            {
-            }
+            public override void WriteByte(byte value) { }
 
-            public override long Seek(long offset, SeekOrigin origin)
-            {
-                return 0;
-            }
+            public override long Seek(long offset, SeekOrigin origin) => 0;
 
-            public override void SetLength(long length)
-            {
-            }
+            public override void SetLength(long length) { }
         }
 
-
-        /// <summary>Used as the IAsyncResult object when using asynchronous IO methods on the base Stream class.</summary>
-        private sealed class SynchronousAsyncResult : IAsyncResult
```

Comment:
The name of this class is making me wonder if it is related to the known `FileStream` async vs sync issues. Is this a correct assumption? Would you mind describing the impact/benefit of removing this class?

Reply:
Nothing to do with FileStream.  This was only used by the NullStream returned by Stream.Null.  BeginRead/Write methods implement the "APM" pattern, which is the original asynchronous pattern in .NET based around the IAsyncResult type.  Every BeginXx method returns an IAsyncResult.  Since all operations on Stream.Null complete synchronously (they're nops), the implementations were returning an instance of this SynchronousAsyncResult type, which was just an "already completed" IAsyncResult.  But we already have one of those we use everywhere else, via the internal TaskToApm type, and so there's no need for the duplication.